### DISCUSSION TITLE
Allow discards in foreach

### DISCUSSION THREAD

**Discussion by adamyakes** [dotnet/csharplang#8948] [Language Ideas]
I was writing some code that randomized some input for testing. I wanted to test it a bunch of times, so I threw it in a loop. Much to my surprise, this doesn't work:

```csharp
foreach (_ in Enumerable.Range(0, 100)
{
    // ...
}
```

I think it would be nice to allow discards here. You can of course still use `var _`, but then `_` becomes a variable and you effectively can't use discards within that block. Note, this silly thing works:

```csharp
foreach (var (_, _) in Enumerable.Range(0, 100).Zip(Enumerable.Range(0, 100)))
{
    // ...
}
```

I recognize this is an incredibly niche use case, and only really applies to this one scenario.

**Comment by TahirAhmadov** [comment]
> ```cs
> foreach (_ in Enumerable.Range(0, 100)
> {
>     // ...
> }
> ```

Why not do:
```cs
for(int i = 0; i < 100; ++i)
{
  // ...
}
```

PS. Or:
```cs
100.Repeat(() =>
{
  // ...
});

public static void Repeat(this int This, Action action) { ... }
```

**Reply by adamyakes** [reply]
The first loop is still introducing a local variable. The second is a little better, but the lambda syntax is ugly IMO and requires an extension method. If we allow _some_ discards in loop variables, what is the restriction to allow all of them?

**Reply by TahirAhmadov** [reply]
What's wrong with a local variable?
The issue with the `foreach (_ in Enumerable.Range(0, 100))` is inefficiency - you're creating an enumerable, there are method calls, etc. Look at this [SharpLab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0AMgJYB2AjgNzpwDMuC2AwtgN7rae71wAs2AWQAUASjYcukgGYB7AE4BTMAGMAFtiEA3MHOwB9bGWwBRUgFcoCuWABGAGwX4ASmFIBzBUMwAabEkyYIiISkpzsaKGRzDKkEDIO+ADqcsTAnnoi1BFRAL4h2HnZnHS4/AIIouJFobJyQmTAhtgAvNiYlE0APH4BHQDUfcTB1ZLhUaFMMXEJyameQ1njhZKFOUA=)

**Comment by TahirAhmadov** [comment]
While it's still an extremely low priority item, I would rather see something like:
```cs
do(100) // any size integer expression goes inside the parens
{
}
```

**Comment by HaloFour** [comment]
See: #7888

**Reply by adamyakes** [reply]
Thanks - I searched for foreach discard and for some reason this did not come up. I've closed this as duplicate.
